File: day07/main.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"log"
	"os"
	"regexp"
	"strconv"
	"strings"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	term := NewTerm()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		if err := term.ReadInput(scanner.Text()); err != nil {
			log.Fatalln(err)
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	totalSize, smallDirs := calcSmallDirSize(term.root)

	fmt.Println("Part 1:", smallDirs)

	freed := 0
	target := minUnused + totalSize - totalDisk
	if target > 0 {
		freed = findDirSizeTarget(term.root, target)
	}
	fmt.Println("Part 2:", freed)
}

const (
	totalDisk = 70000000
	minUnused = 30000000
)

func findDirSizeTarget(n *Node, target int) int {
	if n.size < target {
		return -1
	}
	atMost := n.size
	for _, v := range n.children {
		if !v.isDir {
			continue
		}
		k := findDirSizeTarget(v, target)
		if k >= 0 && k < atMost {
			atMost = k
		}
	}
	return atMost
}

const (
	smallDirLimit = 100000
)

func calcSmallDirSize(n *Node) (int, int) {
	total := 0
	cummulative := 0
	for _, v := range n.children {
		if v.isDir {
			t, c := calcSmallDirSize(v)
			total += t
			cummulative += c
		} else {
			total += v.size
		}
	}
	n.size = total
	if total <= smallDirLimit {
		cummulative += total
	}
	return total, cummulative
}

type (
	Term struct {
		pwd     []string
		root    *Node
		running string
	}

	Node struct {
		name     string
		children map[string]*Node
		isDir    bool
		size     int
	}
)

func NewTerm() *Term {
	return &Term{
		pwd:     nil,
		root:    NewDir("/"),
		running: "",
	}
}

func NewDir(name string) *Node {
	return &Node{
		name:     name,
		children: map[string]*Node{},
		isDir:    true,
		size:     0,
	}
}

func NewFile(name string, size int) *Node {
	return &Node{
		name:     name,
		children: map[string]*Node{},
		isDir:    false,
		size:     size,
	}
}

var (
	cmdRegex = regexp.MustCompile(`^\$ `)
)

func (t *Term) ReadInput(inp string) error {
	if cmdRegex.MatchString(inp) {
		t.running = ""
		return t.Exec(strings.Fields(inp[2:]))
	}
	if t.running != "" {
		switch t.running {
		case "ls":
			return t.ReadOutputLS(inp)
		}
	}
	return errors.New("Invalid input")
}

func (t *Term) Exec(cmd []string) error {
	if len(cmd) == 0 {
		return nil
	}
	switch cmd[0] {
	case "cd":
		{
			if len(cmd) != 2 {
				return errors.New("Invalid cd args")
			}
			return t.CD(cmd[1])
		}
	case "ls":
		{
			if len(cmd) != 1 {
				return errors.New("Invalid ls args")
			}
			t.running = "ls"
			return nil
		}
	default:
		return errors.New("Invalid cmd")
	}
}

func (t *Term) CD(d string) error {
	switch d {
	case "":
		{
			return errors.New("Invalid cd directory")
		}
	case "..":
		{
			if len(t.pwd) == 0 {
				return errors.New("No parent directory from root")
			}
			t.pwd = t.pwd[:len(t.pwd)-1]
			return nil
		}
	case "/":
		{
			t.pwd = t.pwd[:0]
			return nil
		}
	default:
		{
			t.pwd = append(t.pwd, d)
			return nil
		}
	}
}

func (t *Term) ReadOutputLS(inp string) error {
	kind, name, ok := strings.Cut(inp, " ")
	if !ok {
		return errors.New("Invalid ls output")
	}
	if name == "" {
		return errors.New("Invalid ls file name")
	}
	if kind == "dir" {
		return t.Mkdir(name)
	}
	size, err := strconv.Atoi(kind)
	if err != nil {
		return err
	}
	return t.Touch(name, size)
}

func (t *Term) MkdirPath() (*Node, error) {
	node := t.root
	for _, i := range t.pwd {
		if n, ok := node.children[i]; !ok {
			node.children[i] = NewDir(i)
		} else if !n.isDir {
			return nil, errors.New("Mkdir invalid path")
		}
		node = node.children[i]
	}
	return node, nil
}

func (t *Term) Mkdir(name string) error {
	node, err := t.MkdirPath()
	if err != nil {
		return err
	}
	if n, ok := node.children[name]; !ok {
		node.children[name] = NewDir(name)
	} else if !n.isDir {
		return errors.New("Mkdir on non-dir")
	}
	return nil
}

func (t *Term) Touch(name string, size int) error {
	node, err := t.MkdirPath()
	if err != nil {
		return err
	}
	if n, ok := node.children[name]; ok && n.isDir {
		return errors.New("Touch file on dir")
	}
	node.children[name] = NewFile(name, size)
	return nil
}

File: day07/src/main.rs
use lazy_static::lazy_static;
use regex::Regex;
use std::collections::hash_map::Entry;
use std::collections::HashMap;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

type BoxError = Box<dyn std::error::Error>;
type BoxResult<T> = Result<T, BoxError>;

fn main() -> BoxResult<()> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut term = Term::new();

    for line in reader.lines() {
        term.read_input(&line?)?;
    }

    let (total_size, small_dirs) = calc_small_dir_size(&mut term.root);

    let mut freed = 0;
    let target = MIN_UNUSED + total_size - TOTAL_DISK;
    if target > 0 {
        if let Some(k) = find_dir_size_target(&term.root, target) {
            freed = k;
        }
    }

    println!("Part 1: {}\nPart 2: {}", small_dirs, freed);

    Ok(())
}

const TOTAL_DISK: i32 = 70000000;
const MIN_UNUSED: i32 = 30000000;

fn find_dir_size_target(n: &Node, target: i32) -> Option<i32> {
    if n.size < target {
        return None;
    }
    let mut at_most = n.size;
    for v in n.children.values() {
        if !v.is_dir {
            continue;
        }
        if let Some(k) = find_dir_size_target(v, target) {
            if k < at_most {
                at_most = k;
            }
        }
    }
    Some(at_most)
}

const SMALL_DIR_LIMIT: i32 = 100000;

fn calc_small_dir_size(n: &mut Node) -> (i32, i32) {
    let mut total = 0;
    let mut cummulative = 0;
    for v in n.children.values_mut() {
        if v.is_dir {
            let (t, c) = calc_small_dir_size(v);
            total += t;
            cummulative += c;
        } else {
            total += v.size;
        }
    }
    n.size = total;
    if total <= SMALL_DIR_LIMIT {
        cummulative += total;
    }
    (total, cummulative)
}

struct Node {
    children: HashMap<String, Box<Node>>,
    is_dir: bool,
    size: i32,
}

struct Term {
    pwd: Vec<String>,
    root: Box<Node>,
    running: Option<String>,
}

impl Node {
    fn new_dir() -> Box<Self> {
        Box::new(Self {
            children: HashMap::new(),
            is_dir: true,
            size: 0,
        })
    }

    fn new_file(size: i32) -> Box<Self> {
        Box::new(Self {
            children: HashMap::new(),
            is_dir: false,
            size,
        })
    }
}

impl Term {
    fn new() -> Self {
        Self {
            pwd: Vec::new(),
            root: Node::new_dir(),
            running: None,
        }
    }

    fn read_input(&mut self, inp: &str) -> BoxResult<()> {
        lazy_static! {
            static ref RE: Regex = Regex::new(r"^\$ ").unwrap();
        }
        if RE.is_match(inp) {
            self.running = None;
            self.exec(&inp[2..].split_ascii_whitespace().collect::<Vec<_>>()[..])
        } else {
            match &self.running {
                Some(running) => match running.as_str() {
                    "ls" => self.read_output_ls(inp),
                    _ => Err("Invalid input".into()),
                },
                None => Err("Invalid input".into()),
            }
        }
    }

    fn exec(&mut self, cmd: &[&str]) -> BoxResult<()> {
        match cmd {
            &["cd", dir] => self.cd(dir),
            &["ls"] => {
                self.running = Some("ls".into());
                Ok(())
            }
            _ => Err("Invalid cmd".into()),
        }
    }

    fn cd(&mut self, dir: &str) -> BoxResult<()> {
        match dir {
            "" => return Err("No cd dir".into()),
            ".." => {
                self.pwd
                    .pop()
                    .ok_or::<BoxError>("No parent directory from root".into())?;
            }
            "/" => self.pwd.clear(),
            d => self.pwd.push(d.into()),
        }
        Ok(())
    }

    fn read_output_ls(&mut self, inp: &str) -> BoxResult<()> {
        let (kind, name) = if let Some((kind, name)) = inp.split_once(' ') {
            (kind, name)
        } else {
            return Err("Invalid ls output".into());
        };
        if name == "" {
            return Err("Invalid ls file name".into());
        }
        if kind == "dir" {
            self.mkdir(name)
        } else {
            self.touch(name, kind.parse()?)
        }
    }

    fn mkdir_path(&mut self) -> BoxResult<&mut Box<Node>> {
        let mut node = &mut self.root;
        for i in self.pwd.iter() {
            match node.children.entry(i.into()) {
                Entry::Occupied(e) => {
                    let v = e.get();
                    if !v.is_dir {
                        return Err("Mkdir invalid path".into());
                    }
                    node = e.into_mut();
                }
                Entry::Vacant(e) => {
                    node = e.insert(Node::new_dir());
                }
            }
        }
        Ok(node)
    }

    fn mkdir(&mut self, name: &str) -> BoxResult<()> {
        let node = self.mkdir_path()?;
        match node.children.entry(name.into()) {
            Entry::Occupied(e) => {
                if !e.get().is_dir {
                    return Err("Mkdir on non-dir".into());
                }
            }
            Entry::Vacant(e) => {
                e.insert(Node::new_dir());
            }
        }
        Ok(())
    }

    fn touch(&mut self, name: &str, size: i32) -> BoxResult<()> {
        let node = self.mkdir_path()?;
        match node.children.entry(name.into()) {
            Entry::Occupied(mut e) => {
                let v = e.get_mut();
                if v.is_dir {
                    return Err("Touch file on dir".into());
                }
                v.size = size;
            }
            Entry::Vacant(e) => {
                e.insert(Node::new_file(size));
            }
        }
        Ok(())
    }
}
